### PR TITLE
[Feature] Implement Stockpile, Spit Up, Swallow

### PR DESCRIPTION
**Pull request by mcmontag**
## What are the changes?
Implements Stockpile, Spit Up, and Swallow.

## Why am I doing these changes?
Because these moves are interesting.  

The other PRs (https://github.com/pagefaultgames/pokerogue/pull/2954 and https://github.com/pagefaultgames/pokerogue/pull/410) are either partial implementations (the former) or missing some key stat tracking logic.

## What did change?
These implementations rely on a new BattlerTag, named `StockpilingTag`.  The new tag:
- Handles stat changes on adding a stack, or removing all stacks.
- Keeps track of two pieces of state:
  1. The current number of Stockpile stacks
  2. The number of successful changes to DEF and SPDEF (tracked individually) from adding stacks.

In order to support this:
> ii. The number of successful changes to DEF and SPDEF (tracked individually) from adding stacks.

A new (optional) parameter was added to the constructor of `StatChangePhase` -- `onChange: StatChangeCallback`.

```ts
export type StatChangeCallback = (target: Pokemon, changed: BattleStat[], relativeChanges: number[]) => void
```

`changed` contains a list of the `BattleStat`s that will change.
`relativeChanges` is a mapping of `changed` to the relative change in level for each of its (`BattleStat`) entries.

For example, if DEF will increase by 2, and SPDEF will increase by 1, the callback will be invoked like:
```ts
this.onChange(this.getPokemon(), [BattleStat.DEF, BattleStat.SPDEF], [2, 1])
```

`StockpilingTag` provides a callback that increments its internal stat-tracking state for DEF and/or SPDEF if a non-zero value is present in `relativeChanges`.

### Screenshots/Videos

https://github.com/pagefaultgames/pokerogue/assets/54485715/9efc19a8-a06c-430d-899c-577e98f36318

(I got trolled by eggs hatching, no need to watch past the ~1:50 mark)

## How to test the changes?
Most of the simple scenarios are pretty easy to replicate by starting with Cramorant and playing like normal.  Make sure that:
- Spit Up with 1/2/3 stacks has a power of 100/200/300, respectively (use the console)
- Swallow with 1/2/3 heals by 25%/50%/100% (use your eyes)
- Spit Up and Swallow both fail if you have no Stockpile stacks
- Stockpile stacks (and tracked stat changes) carry over as expected between wild pokemon encounters

Overrides and/or additional log messages are likely the easiest way to test edge cases. For example:
- Simple/Contrary interactions
- Stockpile should work (increment stacks if possible) even if both DEF and SPDEF are at max
- Not counting (per stat) stat changes for any stacks that did not raise DEF and/or SPDEF due to max thresholds

I'm looking into the testing framework; if automated tests don't make it into this PR, I'll add them as a fast-follow.

## Checklist
- [x] There is no overlap with another PR? (mentioned above)
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes? (open to adding more if need be)

### CODE REVIEW

File: src/locales/ko/battle.ts [hunk 1]
```diff
@@ -135,4 +135,5 @@ export const battle: SimpleTranslationEntries = {
   "battlerTagsSaltCuredLapse": "{{pokemonNameWithAffix}}[[는]] 소금절이의\n데미지를 입고 있다.",
   "battlerTagsCursedOnAdd": "{{pokemonNameWithAffix}}[[는]]\n자신의 체력을 깎아서\n{{pokemonName}}에게 저주를 걸었다!",
   "battlerTagsCursedLapse": "{{pokemonNameWithAffix}}[[는]]\n저주받고 있다!",
+  "battlerTagsStockpilingOnAdd": "{{pokemonNameWithAffix}} stockpiled {{stockpiledCount}}!"
```
Comment (EnochG1) [inline src/locales/ko/battle.ts:138]:
```suggestion
  "battlerTagsStockpilingOnAdd": "{{pokemonNameWithAffix}}[[는]]\n{{stockpiledCount}}개 비축했다!",
```

**Comment by flx-sta** [comment]
Pull the latest beta branch to fix the failing tests

**Comment by mcmontag** [comment]
I may have done the rebase incorrectly...not sure why it's showing things like `src/test/moves/dynamax_cannon.test.ts` as changed files when they're already in beta.

**Comment by torranx** [comment]
Can we add integration tests?

**Comment by mcmontag** [comment]
For sure - I'll do that when I get a few minutes today!

**Comment by mcmontag** [comment]
@torranx I've added tests 👍

File: src/data/move.ts [hunk 2]
```diff
@@ -6483,12 +6540,17 @@ export function initMoves() {
       .target(MoveTarget.RANDOM_NEAR_ENEMY)
       .partial(),
     new SelfStatusMove(Moves.STOCKPILE, Type.NORMAL, -1, 20, -1, 0, 3)
-      .unimplemented(),
-    new AttackMove(Moves.SPIT_UP, Type.NORMAL, MoveCategory.SPECIAL, -1, 100, 10, -1, 0, 3)
-      .unimplemented(),
+      .condition(user => ((user.getTag(BattlerTagType.STOCKPILING) as StockpilingTag)?.stockpiledCount ?? 0) < 3)
```
Comment (mcmontag) [inline src/data/move.ts:6543]:
```suggestion
      .condition(user => user.getTag(StockpilingTag)?.stockpiledCount < 3)
```

File: src/data/move.ts [hunk 3]
```diff
@@ -6483,12 +6540,17 @@ export function initMoves() {
       .target(MoveTarget.RANDOM_NEAR_ENEMY)
       .partial(),
     new SelfStatusMove(Moves.STOCKPILE, Type.NORMAL, -1, 20, -1, 0, 3)
-      .unimplemented(),
-    new AttackMove(Moves.SPIT_UP, Type.NORMAL, MoveCategory.SPECIAL, -1, 100, 10, -1, 0, 3)
-      .unimplemented(),
+      .condition(user => user.getTag(StockpilingTag)?.stockpiledCount < 3)
```
Comment (mcmontag) [inline src/data/move.ts:6543]:
```suggestion
      .condition(user => (user.getTag(StockpilingTag)?.stockpiledCount ?? 0) < 3)
```

**Review by torranx — APPROVED** [review]
lgtm

**Pull request merged**